File: src/tests/image_tests.cpp
#include <test.hpp>

#include <core\core.hpp>
#include <image\image.hpp>

using namespace image;

SUITE(ImageTests) {
  TEST(CanCreate) {
    shared_ptr<Image> im(new Image(48, 32));
    
    CHECK_EQUAL(48, im->getWidth());
    CHECK_EQUAL(32, im->getHeight());
  }

  TEST(CanLoad) {
    {
      shared_ptr<Image> im(new Image("..\\res\\tests\\test.png"));
      
      CHECK_EQUAL(16, im->getWidth());
      CHECK_EQUAL(16, im->getHeight());
    }

    {
      shared_ptr<Image> im(new Image("..\\res\\tests\\test.jpg"));
      
      CHECK_EQUAL(96, im->getWidth());
      CHECK_EQUAL(96, im->getHeight());
    }

    {
      shared_ptr<Image> im(new Image("..\\res\\tests\\test.gif"));
      
      CHECK_EQUAL(80, im->getWidth());
      CHECK_EQUAL(40, im->getHeight());
    }
  }
}

SUITE(ImageListTests) {
  TEST(CanCreate) {
    shared_ptr<ImageList> il(new ImageList());
  }
}

File: src/tests/lua/image_tests.lua
function test_load()
    im = Image("..\\res\\tests\\test.jpg")
    
    checkEqual(96, im.width)
    checkEqual(96, im.height)
end

function test_create()
    im = Image(32, 32)
    
    checkEqual(32, im.width)
    checkEqual(32, im.height)
end

function test_getPixel()
    im = Image("..\\res\\tests\\test.jpg")
    
    r, g, b, a = im:getPixel(0, 0)
    checkEqual({213, 152, 219, 255}, {r, g, b, a})
    r, g, b, a = im:getPixel(16, 16)
    checkEqual({216, 157, 205, 255}, {r, g, b, a})
    r, g, b, a = im:getPixel(32, 32)
    checkEqual({16, 13, 34, 255}, {r, g, b, a})
end

function test_setPixel()
    im = Image(4, 4)
    
    r, g, b, a = im:getPixel(0, 0)
    checkEqual({0, 0, 0, 0}, {r, g, b, a})

    im:setPixel(0, 0, 1, 2, 3, 4)
    r, g, b, a = im:getPixel(0, 0)
    checkEqual({1, 2, 3, 4}, {r, g, b, a})

    im:setPixel(1, 1, {2, 4, 6, 8})
    r, g, b, a = im:getPixel(1, 1)
    checkEqual({2, 4, 6, 8}, {r, g, b, a})
end

function test_save()
    im = Image(2, 2)
    
    im:setPixel(0, 0, 255, 0, 0, 255)
    im:setPixel(1, 0, 0, 255, 0, 255)
    im:setPixel(0, 1, 0, 0, 255, 255)
    im:setPixel(1, 1, 255, 255, 0, 255)
    
    check(im:save("test.png"))
    
    im = Image("test.png")

    r, g, b, a = im:getPixel(0, 0)
    checkEqual({255, 0, 0, 255}, {r, g, b, a})
    r, g, b, a = im:getPixel(1, 0)
    checkEqual({0, 255, 0, 255}, {r, g, b, a})
    r, g, b, a = im:getPixel(0, 1)
    checkEqual({0, 0, 255, 255}, {r, g, b, a})
    r, g, b, a = im:getPixel(1, 1)
    checkEqual({255, 255, 0, 255}, {r, g, b, a})
    
    os.remove("test.png")
end
